File: DataModuleTester/src/com/mobnetic/coinguardiandatamodule/tester/volley/CheckerVolleyMainRequest.java
package com.mobnetic.coinguardiandatamodule.tester.volley;

import android.text.TextUtils;

import com.android.volley.DefaultRetryPolicy;
import com.android.volley.Response.ErrorListener;
import com.android.volley.Response.Listener;
import com.android.volley.toolbox.RequestFuture;
import com.mobnetic.coinguardian.model.CheckerInfo;
import com.mobnetic.coinguardian.model.Market;
import com.mobnetic.coinguardian.model.Ticker;
import com.mobnetic.coinguardiandatamodule.tester.volley.CheckerVolleyMainRequest.TickerWithRawResponse;
import com.mobnetic.coinguardiandatamodule.tester.volley.generic.GenericCheckerVolleyRequest;

public class CheckerVolleyMainRequest extends GenericCheckerVolleyRequest<TickerWithRawResponse> {
	
	private final Market market;

	public CheckerVolleyMainRequest(Market market, CheckerInfo checkerInfo, Listener<TickerWithRawResponse> listener, ErrorListener errorListener) {
		super(market.getUrl(0, checkerInfo), checkerInfo, listener, errorListener);
		setRetryPolicy(new DefaultRetryPolicy(5000, 3, 1.5f));
		this.market = market;
	}
	
	@Override
	protected TickerWithRawResponse parseNetworkResponse(String responseString) throws Exception {
		TickerWithRawResponse tickerWithRawResponse = new TickerWithRawResponse();
		tickerWithRawResponse.rawResponse = responseString;
		try {
			tickerWithRawResponse.ticker = market.parseTickerMain(0, responseString, new Ticker(), checkerInfo);
		} catch (Exception e) {
			e.printStackTrace();
			tickerWithRawResponse.ticker = null;
		}
		
		if(tickerWithRawResponse.ticker==null || tickerWithRawResponse.ticker.last<=Ticker.NO_DATA) {
			String errorMsg;
			try {
				errorMsg = market.parseErrorMain(0, responseString, checkerInfo);
			} catch (Exception e) {
				errorMsg = null;
			}
			throw new CheckerErrorParsedError(responseString, errorMsg);
		}
		
		final int numOfRequests = market.getNumOfRequests(checkerInfo);
		if(numOfRequests>1) {
			for(int requestId=1; requestId<numOfRequests; ++requestId) {
				try {
					RequestFuture<String> future = RequestFuture.newFuture();
					final String nextUrl = market.getUrl(requestId, checkerInfo);
					if(!TextUtils.isEmpty(nextUrl)) {
						CheckerVolleyNextRequest request = new CheckerVolleyNextRequest(nextUrl, checkerInfo, future);
						getRequestQueue().add(request);
						String nextResponse = future.get(); // this will block
						market.parseTickerMain(requestId, nextResponse, tickerWithRawResponse.ticker, checkerInfo);
					}
				} catch (Exception e) {
					e.printStackTrace();
				}
			}
		}
		return tickerWithRawResponse;
	}
	
	public class TickerWithRawResponse {
		
		public Ticker ticker;
		public String rawResponse;
		
	}
}

File: dataModuleTester/src/main/java/com/aneonex/bitcoinchecker/tester/volley/CheckerVolleyMainRequest.kt
package com.aneonex.bitcoinchecker.tester.volley

import android.text.TextUtils
import com.android.volley.DefaultRetryPolicy
import com.android.volley.Response
import com.android.volley.toolbox.RequestFuture
import com.aneonex.bitcoinchecker.datamodule.model.CheckerInfo
import com.aneonex.bitcoinchecker.datamodule.model.Market
import com.aneonex.bitcoinchecker.datamodule.model.Ticker
import com.aneonex.bitcoinchecker.tester.volley.CheckerVolleyMainRequest.TickerWrapper
import com.aneonex.bitcoinchecker.tester.volley.generic.GenericCheckerVolleyRequest

class CheckerVolleyMainRequest(market: Market, checkerInfo: CheckerInfo, listener: Response.Listener<TickerWrapper?>, errorListener: Response.ErrorListener)
    : GenericCheckerVolleyRequest<TickerWrapper?>(market.getUrl(0, checkerInfo), checkerInfo, listener, errorListener) {

    private val market: Market
    @Throws(Exception::class)
    override fun parseNetworkResponse(headers: Map<String?, String?>?, responseString: String?): TickerWrapper? {
        val tickerWrapper = TickerWrapper()
        try {
            tickerWrapper.ticker = market.parseTickerMain(0, responseString!!, Ticker(), checkerInfo)
        } catch (e: Exception) {
            e.printStackTrace()
            tickerWrapper.ticker = null
        }
        if (tickerWrapper.ticker == null || tickerWrapper.ticker!!.last <= Ticker.NO_DATA) {
            val errorMsg: String? = try {
                market.parseErrorMain(0, responseString!!, checkerInfo)
            } catch (e: Exception) {
                null
            }
            throw CheckerErrorParsedError(errorMsg)
        }

        val numOfRequests = market.getNumOfRequests(checkerInfo)
        if (numOfRequests > 1) {
            for (requestId in 1 until numOfRequests) {
                try {
                    val future = RequestFuture.newFuture<String>()
                    val nextUrl = market.getUrl(requestId, checkerInfo)
                    if (!TextUtils.isEmpty(nextUrl)) {
                        val request = CheckerVolleyNextRequest(nextUrl, checkerInfo, future)
                        requestQueue!!.add(request)
                        val nextResponse = future.get() // this will block
                        market.parseTickerMain(requestId, nextResponse, tickerWrapper.ticker!!, checkerInfo)
                    }
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }
        }
        return tickerWrapper
    }

    inner class TickerWrapper {
        var ticker: Ticker? = null
    }

    init {
        retryPolicy = DefaultRetryPolicy(5000, 3, 1.5f)
        this.market = market
    }
}